src/cart/crt_rpc.c

```diff
@@ -1532,7 +1532,7 @@ crt_req_send(crt_rpc_t *req, crt_cb_t complete_cb, void *arg)
 			/* failure already reported through complete cb */
 			if (complete_cb != NULL)
 				rc = 0;
-		} else {
+		} else if (!crt_rpc_completed(rpc_priv)) {
```

Comment:
Copying here the feedback from 2.6 branch:

Can you explain how an rpc can be validly completed at this point?

From my view, to complete, we would need to send it via 1515 rc = crt_req_send_internal(rpc_priv); , but that call should either a return rc on error, or send the rpc and invoke a completion cb. Your check here implies that somehow both can happen.

One note to make here, we only used crt_rpc_completed() to detect bugs in the ref counting logic or elsewhere before. I don't think we should start using this function to decide ref counting, and instead should fix underlying ref count problem instead

Reply:
as discussed in 2.6 pr, this is fine for now. cart team will refine rpc completion logic later.

Reply:
I did some investigation about the corpc reference, pushed another small PR -
https://github.com/daos-stack/daos/pull/15572
The refcount possibly can be fourthly refined later, the PR15572 just try to refine PR15476. 

@frostedcmos the refcount handling in crt_corpc_complete() looks correct, see below debug log for the refcount trace for the root node of a bcast, the same last number in the "()" means the corresponding addref and decref. The refcount handling looks fine, but the CRT_RPC_FLAG_CO_FAILOUT fail handling looks caused the issue.

12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:446 crt_corpc_req_create() corpc 0x7f329c46a740 created
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:783 crt_req_addref() rank 0 rpc 0x7f329c46a740, addref to 2                   (1) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:1482 crt_req_send() rank 0 rpc 0x7f329c46a740, addref to 3                    (2) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:787 crt_corpc_req_hdlr() rank 0 rpc 0x7f329c46a740, addref to 4            (3) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:889 crt_corpc_req_hdlr() rank 0 rpc 0x7f329c46a740, addref to 5            (4) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:889 crt_corpc_req_hdlr() rank 0 rpc 0x7f329c46a740, addref to 6            (5) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:1854 crt_rpc_common_hdlr() rank 0 rpc 0x7f329c46a740, addref to 7             (6) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:1544 crt_req_send() rank 0 rpc 0x7f329c46a740, decref to 6                    (2) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/74] rpc  ERR  src/cart/crt_rpc.c:1789 crt_handle_rpc() rank 0 rpc 0x7f329c46a740, addref to 7                  (7) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/74] rpc  ERR  src/cart/crt_rpc.c:1798 crt_handle_rpc() rank 0 rpc 0x7f329c46a740, decref to 6                  (7) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/74] rpc  ERR  src/cart/crt_rpc.c:1804 crt_handle_rpc() rank 0 rpc 0x7f329c46a740, decref to 5                  (6) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:768 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, decref to 4           (4) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:755 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, addref to 5           (8) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] rpc  ERR  src/cart/crt_context.c:550 crt_rpc_complete_and_unlock() rank 0 rpc 0x7f329c46a740, decref to 4 (creat)
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:576 crt_corpc_complete() rank 0 rpc 0x7f329c46a740, decref to 3             (3) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:763 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, decref to 2           (8) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:768 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, decref to 1           (5) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:789 crt_req_decref() rank 0 rpc 0x7f329c46a740, decref to 0                   (1)